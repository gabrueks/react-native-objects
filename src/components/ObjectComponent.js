import React, {useState} from 'react';
import {
  View,
  ImageBackground,
  StyleSheet,
  TouchableHighlight,
} from 'react-native';
import generateRandomColorOrImage from '../helpers/generateRandomColorOrImage';

let taps = 0;

export default ({styleFunction, paramStyles}) => {
  const [color, setColor] = useState(
    paramStyles.borderBottomColor || paramStyles.backgroundColor,
  );
  const [uri, setUri] = useState(paramStyles.uri);
  const zIndex = 9999;
  const clickOnObject = () => {
    taps++;
    setTimeout(async () => {
      if (taps === 1) {
        taps = 0;
      } else if (taps === 2) {
        if (color) {
          taps = 0;
          console.log('s');
          setColor(
            (await generateRandomColorOrImage({circle: true})).backgroundColor,
          );
        } else {
          taps = 0;
          setUri((await generateRandomColorOrImage({square: true})).uri);
        }
      }
    }, 300);
  };

  if (paramStyles.borderStyle && paramStyles.uri) {
    const flexDirection = 'row';

    return (
      <TouchableHighlight
        onPress={clickOnObject}
        style={{
          ...styleFunction({
            ...paramStyles,
            borderBottomColor: color,
            backgroundColor: paramStyles.borderBottomColor
              ? 'transparent'
              : color,
          }),
          zIndex,
        }}>
        <ImageBackground
          source={{uri}}
          key={Date.now()}
          imageStyle={{
            borderRadius: paramStyles.borderRadius,
            borderLeftWidth: paramStyles.borderLeftWidth,
            borderRightWidth: paramStyles.borderRightWidth,
            borderBottomWidth: paramStyles.borderBottomWidth,
            width: paramStyles.width,
            height: paramStyles.height,
          }}
          style={{width: paramStyles.width, height: paramStyles.height}}>
          <View style={{flexDirection}}>
            <View
              style={styles.triangleFirstView(paramStyles.borderRightWidth)}
            />
            <View
              style={styles.triangleSecondView(paramStyles.borderRightWidth)}
            />
          </View>
        </ImageBackground>
      </TouchableHighlight>
    );
  }
  if (paramStyles.uri) {
    return (
      <TouchableHighlight
        onPress={clickOnObject}
        style={{
          ...styleFunction(paramStyles),
          zIndex,
        }}>
        <ImageBackground
          source={{uri}}
          key={Date.now()}
          imageStyle={{
            borderLeftWidth: paramStyles.borderLeftWidth,
            borderRightWidth: paramStyles.borderRightWidth,
            borderBottomWidth: paramStyles.borderBottomWidth,
            width: paramStyles.width,
            height: paramStyles.height,
          }}
          style={{width: paramStyles.width, height: paramStyles.height}}
        />
      </TouchableHighlight>
    );
  }

  return (
    <TouchableHighlight
      onPress={clickOnObject}
      style={{
        ...styleFunction({
          ...paramStyles,
          borderBottomColor: paramStyles.borderBottomColor ? color : null,
          backgroundColor: paramStyles.borderBottomColor
            ? 'transparent'
            : color,
        }),
        zIndex,
      }}>
      <View />
    </TouchableHighlight>
  );
};

const styles = StyleSheet.create({
  triangleFirstView: width => ({
    width: 0,
    height: 0,
    backgroundColor: 'transparent',
    borderStyle: 'solid',
    borderRightWidth: width,
    borderTopWidth: width * 2,
    borderRightColor: 'transparent',
    borderTopColor: '#FFF',
  }),
  triangleSecondView: width => ({
    width: 0,
    height: 0,
    backgroundColor: 'transparent',
    borderStyle: 'solid',
    borderLeftWidth: width,
    borderTopWidth: width * 2,
    borderLeftColor: 'transparent',
    borderTopColor: '#FFF',
  }),
});
