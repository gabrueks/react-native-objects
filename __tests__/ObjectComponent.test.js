import React from 'react';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';

import ObjectComponent from '../src/components/ObjectComponent';
import timeout from '../src/helpers/timeout';

const createCircleStyle = ({
  width,
  height,
  borderRadius,
  backgroundColor,
  left,
  top,
}) => ({
  width,
  height,
  backgroundColor,
  left,
  top,
  borderRadius,
  position: 'absolute',
});

const createSquareStyle = ({width, height, backgroundColor, left, top}) => ({
  width,
  height,
  backgroundColor,
  left,
  top,
  position: 'absolute',
});

const createTrianglestyle = ({
  width,
  height,
  backgroundColor,
  left,
  top,
  borderLeftWidth,
  borderRightWidth,
  borderBottomWidth,
  borderLeftColor,
  borderRightColor,
  borderBottomColor,
  borderStyle,
}) => ({
  width,
  height,
  backgroundColor,
  left,
  top,
  borderLeftWidth,
  borderRightWidth,
  borderBottomWidth,
  borderLeftColor,
  borderRightColor,
  borderStyle,
  borderBottomColor,
  position: 'absolute',
});

const triangleStyles = {
  backgroundColor: 'transparent',
  borderBottomColor: '#6BA0D7',
  borderBottomWidth: 127,
  borderLeftColor: 'transparent',
  borderLeftWidth: 63.5,
  borderRightColor: 'transparent',
  borderRightWidth: 63.5,
  borderStyle: 'solid',
  height: 0,
  left: 164,
  top: 435,
  width: 0,
};

const circleStyles = {
  backgroundColor: '#252450',
  borderRadius: 35,
  height: 70,
  left: 107,
  top: 250,
  width: 70,
};

const squareStyles = {
  height: 131,
  left: 101.5,
  top: 396.5,
  uri: 'http://static.colourlovers.com/images/patterns/1978/1978162.png',
  width: 131,
};

const circleComponentWrapper = () => (
  <ObjectComponent
    styleFunction={createCircleStyle}
    paramStyles={circleStyles}
  />
);

const squareComponentWrapper = () => (
  <ObjectComponent
    styleFunction={createSquareStyle}
    paramStyles={squareStyles}
  />
);

const triangleComponentWrapper = () => (
  <ObjectComponent
    styleFunction={createTrianglestyle}
    paramStyles={triangleStyles}
  />
);

describe('Testing object component', () => {
  it('Should create a circle', async () => {
    const instance = renderer.create(circleComponentWrapper()).toJSON();
    expect(instance).toMatchSnapshot();
  });

  it('Should call the onPress function and change color of the circle.', async () => {
    const instance = shallow(circleComponentWrapper());
    instance
      .find('TouchableHighlight')
      .first()
      .prop('onPress')();
    instance
      .find('TouchableHighlight')
      .first()
      .prop('onPress')();
    await timeout(2000);
    const newBackgroundColor = instance
      .find('TouchableHighlight')
      .first()
      .props().style.backgroundColor;

    expect(newBackgroundColor === circleStyles.backgroundColor).toBeFalsy();
  });

  it('Should create a square', async () => {
    const instance = renderer.create(squareComponentWrapper()).toJSON();
    expect(instance).toMatchSnapshot();
  });

  it('Should call the onPress function and change uri of the square.', async () => {
    const instance = shallow(squareComponentWrapper());
    instance
      .find('TouchableHighlight')
      .first()
      .prop('onPress')();
    instance
      .find('TouchableHighlight')
      .first()
      .prop('onPress')();
    await timeout(2000);
    const newUri = instance
      .find('ImageBackground')
      .first()
      .props().style.uri;

    expect(newUri === squareStyles.uri).toBeFalsy();
  });

  it('Should create a triangle', async () => {
    const instance = renderer.create(triangleComponentWrapper()).toJSON();
    expect(instance).toMatchSnapshot();
  });

  it('Should call the onPress function and change color of the triangle.', async () => {
    const instance = shallow(triangleComponentWrapper());
    instance
      .find('TouchableHighlight')
      .first()
      .prop('onPress')();
    instance
      .find('TouchableHighlight')
      .first()
      .prop('onPress')();
    await timeout(2000);
    const newColor = instance
      .find('TouchableHighlight')
      .first()
      .props().style.borderBottomColor;

    expect(newColor === triangleStyles.borderBottomColor).toBeFalsy();
  });
});
