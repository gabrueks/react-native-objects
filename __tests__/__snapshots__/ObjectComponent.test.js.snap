// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing object component Should create a circle 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#252450",
      "borderRadius": 35,
      "height": 70,
      "left": 107,
      "position": "absolute",
      "top": 250,
      "width": 70,
      "zIndex": 9999,
    }
  }
>
  <View
    style={null}
  />
</View>
`;

exports[`Testing object component Should create a square 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": undefined,
      "height": 131,
      "left": 101.5,
      "position": "absolute",
      "top": 396.5,
      "width": 131,
      "zIndex": 9999,
    }
  }
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      Object {
        "height": 131,
        "width": 131,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "http://static.colourlovers.com/images/patterns/1978/1978162.png",
        }
      }
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": 131,
            "width": 131,
          },
          Object {
            "borderBottomWidth": undefined,
            "borderLeftWidth": undefined,
            "borderRightWidth": undefined,
            "height": 131,
            "width": 131,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`Testing object component Should create a triangle 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "transparent",
      "borderBottomColor": "#6BA0D7",
      "borderBottomWidth": 127,
      "borderLeftColor": "transparent",
      "borderLeftWidth": 63.5,
      "borderRightColor": "transparent",
      "borderRightWidth": 63.5,
      "borderStyle": "solid",
      "height": 0,
      "left": 164,
      "position": "absolute",
      "top": 435,
      "width": 0,
      "zIndex": 9999,
    }
  }
>
  <View
    style={null}
  />
</View>
`;
